sphericart/include/dynamic_cuda.hpp

```diff
@@ -0,0 +1,306 @@
+    } while (0)
+
+#define CUDART_SAFE_CALL(call)                                                                     \
+    do {                                                                                           \
+        cudaError_t cudaStatus = (call);                                                           \
+        if (cudaStatus != cudaSuccess) {                                                           \
+            std::cerr << "CUDA error at " << __FILE__ << ":" << __LINE__ << " - "                  \
+                      << CUDART::instance().cudaGetErrorString(cudaStatus) << '\n'                 \
+                      << "File: " << __FILE__ << '\n'                                              \
+                      << "Line: " << __LINE__ << '\n';                                             \
+            exit(1);                                                                               \
+        }                                                                                          \
+    } while (0)
+
+// Define a template to dynamically load symbols
+template <typename FuncType> FuncType load(void* handle, const char* functionName) {
+    auto func = reinterpret_cast<FuncType>(dlsym(handle, functionName));
+    if (!func) {
+        throw std::runtime_error(std::string("Failed to load function: ") + functionName);
+    }
+    return func;
+}
+
+class CUDART {
+  public:
+    static CUDART& instance() {
+        static CUDART instance;
+        return instance;
+    }
+
+    using cudaGetDeviceCount_t = cudaError_t (*)(int*);
+    using cudaGetDevice_t = cudaError_t (*)(int*);
+    using cudaSetDevice_t = cudaError_t (*)(int);
+    using cudaMalloc_t = cudaError_t (*)(void**, size_t);
+    using cudaMemcpy_t = cudaError_t (*)(void*, const void*, size_t, cudaMemcpyKind);
+    using cudaGetErrorName_t = const char* (*)(cudaError_t);
+    using cudaGetErrorString_t = const char* (*)(cudaError_t);
+    using cudaDeviceSynchronize_t = cudaError_t (*)(void);
+    using cudaPointerGetAttributes_t = cudaError_t (*)(cudaPointerAttributes*, const void*);
+    using cudaFree_t = cudaError_t (*)(void*);
+
+    cudaGetDeviceCount_t cudaGetDeviceCount;
+    cudaGetDevice_t cudaGetDevice;
+    cudaSetDevice_t cudaSetDevice;
+    cudaMalloc_t cudaMalloc;
+    cudaMemcpy_t cudaMemcpy;
+    cudaGetErrorName_t cudaGetErrorName;
+    cudaGetErrorString_t cudaGetErrorString;
+    cudaDeviceSynchronize_t cudaDeviceSynchronize;
+    cudaPointerGetAttributes_t cudaPointerGetAttributes;
+    cudaFree_t cudaFree;
+
+  private:
+    CUDART() {
+        cudartHandle = dlopen("libcudart.so", RTLD_NOW);
+
+        if (!cudartHandle) {
+            throw std::runtime_error(
+                "Failed to load libcudart.so. Try running \"find /usr -name libcudart.so\" and "
+                "appending the directory to your $LD_LIBRARY_PATH environment variable."
+            );
```

Comment:
```suggestion
            throw std::runtime_error(
                "Failed to load libcudart.so. Make sure it is available in your $LD_LIBRARY_PATH environment variable."
            );
```

The `find /usr` suggestion might be wrong on HPC, where cuda might be in modules which don't live in `/usr`